### DISCUSSION TITLE
Headers Size

### DISCUSSION THREAD

**Discussion by IlyaDjurin** [Kludex/uvicorn#1669] [Q&A]
In uvicorn options  i can't see something similar to "--limit-request-header-size" (gunicorn).  I need this option, because i have request with headers(cookies)  the size of which more then 4096 bytes, and uvicorn raise exception(uvicorn.error) in FastAPI(websocket). When size less then 4096 - request ok. Suggest to me, is there  solution to this issue ?

**Comment by Kludex** [comment]
The solution would be to implement that on uvicorn itself. 😅 

Ref.: https://github.com/encode/uvicorn/issues/157

**Reply by euri10** [reply]
what exception ?
shouldn't that be dealt upstream ?

**Reply by IlyaDjurin** [reply]
exception only in log
`
INFO:     connection failed (400 Bad Request)
{"thread": 12, "lineno": 225, "levelname": "INFO", "msg": "connection failed (400 Bad Request)", "name": "uvicorn.error", "asctime": "2022-09-27 12:26:27+00:00", "app_label": "SOCKET-SERVICE", "duration": 596, "payload": {}, "exception": {}, "query_params": {}, "response": {}, "status_code": null}
`

**Reply by iudeen** [reply]
@IlyaDjurin did you try this using wsproto? 

```
uvicorn app:app --ws wsproto
```

Adding more context:

I tried to replicate this scenario (was trying to get started with implementing `--limit-request-header-size`):

Code used to test:
```python
from websocket import create_connection

GET_REQUEST_HUGE_HEADERS = [
    "".join(
        [
            "Cookie: " + "x" * 32 * 4096,
        ]
    )
]

ws = create_connection("ws://localhost:8000/ws", header=GET_REQUEST_HUGE_HEADERS)
print("Sending 'Hello, World'...")
ws.send("Hello, World")
print("Sent")
print("Receiving...")
result = ws.recv()
print("Received '%s'" % result)
```

The above code fails for Websockets implementation, but works with wsproto implementation.

**Reply by iudeen** [reply]
Looking at websockets code (`websockets\legacy\http.py`) line 195, we see 

https://github.com/aaugustin/websockets/blob/569ba1e04aa78313fd4bceb1cdae59c822682add/src/websockets/legacy/http.py#L196  

```python
    # Security: this guarantees header values are small (hard-coded = 4 KiB)
    if len(line) > MAX_LINE:
        raise SecurityError("line too long")
```

**Comment by plmatev** [comment]
This [commit](https://github.com/python-websockets/websockets/commit/bbb316155a5aeb719f262873a5b29a98c19b25d9) in websockets implementation should give control on the line size. Now it is reading environment variable `WEBSOCKETS_MAX_LINE_LENGTH` (with default value of 8192 bytes)

[Implementation](https://github.com/python-websockets/websockets/blob/bbb316155a5aeb719f262873a5b29a98c19b25d9/src/websockets/legacy/http.py)

```python
MAX_LINE_LENGTH = int(os.environ.get("WEBSOCKETS_MAX_LINE_LENGTH", "8192"))
```

```python
    # Security: this guarantees header values are small (hard-coded = 8 KiB)
    if len(line) > MAX_LINE_LENGTH:
        raise SecurityError("line too long")
```
